src/test/java/com/google/graphgeckos/dashboard/storage/DatastoreRepositoryTests.java

```diff
@@ -0,0 +1,220 @@
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class DatastoreRepositoryTests {
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+
+  private BuildInfo getDummyEntity(String commitHash) {
+    return new BuildInfo(new ParsedGitData(commitHash, Timestamp.now(), "test"));
+  }
+
+  private ParsedBuildbotData getDummyUpdate(String commitHash) {
+    return new ParsedBuildbotData(commitHash, "tester", new ArrayList<>(), BuilderStatus.PASSED);
+  }
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void testValidAddition() {
+    DatastoreRepository storage = new DatastoreRepository();
+    
+    Assert.assertTrue(storage.createRevisionEntry(getDummyEntity("1")));
+    Assert.assertEquals(getDummyEntity("1"), storage.getRevisionEntry("1"));
+  }
+
+  @Test
+  public void testValidUpdate() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertTrue(storage.updateRevisionEntry(getDummyUpdate("1")));
+    BuildInfo dummy = getDummyEntity("1");
+    dummy.addBuilder(new Builder(getDummyUpdate("1")));
+    Assert.assertEquals(dummy, storage.getRevisionEntry("1"));
+  }
+
+  @Test
+  public void testValidDeletion() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertTrue(storage.deleteRevisionEntry("1"));
+    Assert.assertNull(storage.getRevisionEntry("1"));
+  }
+
+  /**
+   * Tests the behaviour of add/update/get/delete requests on null data.
+   */
+  @Test
+  public void testRequestsNullData() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertFalse(storage.createRevisionEntry(null));
+    Assert.assertFalse(storage.getRevisionEntry(null));
+    Assert.assertFalse(storage.updateRevisionEntry(null));
+    Assert.assertFalse(storage.deleteRevisionEntry(null));
+  }
+
+  @Test
+  public void testAddingSameEntry() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertTrue(storage.createRevisionEntry(getDummyEntity("1")));
+    Assert.assertFalse(storage.createRevisionEntry(getDummyEntity("1")));
+  }
+
+  @Test
+  public void testUpdatingInexistentEntry() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertFalse(storage.updateRevisionEntry(getDummyUpdate("2")));
+  }
+
+  @Test
+  public void testGettingInexistentEntry() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertNull(storage.getRevisionEntry("2"));
+  }
+
+  @Test
+  public void testDeletingInexistentEntry() {
+    DatastoreRepository storage = new DatastoreRepository();
+
+    Assert.assertTrue(storage.deleteRevisionEntry("2"));
+  }
+
+  /**
+   * Tests multiple threads creating update requests for the same entities.
+   */
+  @Test
+  public void testParallelBurstRequests() {
```

Comment:
I would not include a multithreaded unit test.
Unit tests like this are not the best way to test or set concurrency expectations. We have no control over how the Threads are going to be scheduled and how are they to interfere with each other, so effectively every run of the test could execute a very different execution. Often such tests become "flaky" (different executions pass or fail in the same configuration) which reduces their utility as a good way to prevent regressions.

Reply:
How can we test concurrent use of the datastore then?

Reply:
>How can we test concurrent use of the datastore then?

We don't.
We rely on the way datastore handles concurrent use itself. Concurrency guarantees are properties of the code that are not well exercised by traditional testing. Typically people would review the code and check if it acts within the specifications. In a more formal setting, it's possible to employ some formal proof techniques to have a mechanically verified proof of correctness (stuff like [TLA+](https://en.wikipedia.org/wiki/TLA%2B)), but that's not practical to do at the application code level as it's a lot of work. So usually those methods are employed for the code that actually implements the database layer, and the application layer just assumes well-behaved concurrency behavior from the database. In particular, Datastore enforces [serializable isolation](https://cloud.google.com/datastore/docs/concepts/transactions#isolation_and_consistency) and the code here should be well-behaved in such context.

Reply:
> Datastore enforces serializable isolation and the code here should be well-behaved in such context.

I believe this applies to Datastore transactions, which are a separate part of the Datastore API, since there is no mention of race conditions for regular operations, like add, delete or run. I have deleted this test for now, but I think it's worth testing if everything runs as expected in a parallel environment.